### PR TITLE
Rework DefinesClass lookup.

### PR DESCRIPTION
**Pull request by gkossakowski**
Before this commit, DefinesClass meant two things:
1. Function String => Boolean that determined whether a class with
    a given binary name is defined by a classpath entry the function
    instance is associated with
2. A function File => (String => Boolean) that created function from 1.
    for a given classpath entry

After this commit, `DefinesClass` means only the function from 1.

The operation from 2. is moved to newly introduced PerClasspathEntryLookup
interface. That interface also includes lookup of Analysis that corresponds
to a given classpath entry.

Thanks to this change, conversions between Java-defined function-like
objects and Scala functions is largely removed. These conversions were
error prone as shown by #100. As a bonus, the refactoring from this commit
makes it a bit easier to implement sbt/sbt#2525.

### CODE REVIEW

File: internal/compiler-interface/src/main/java/xsbti/compile/DefinesClass.java [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+ * Determines if a classpath entry contains a class.
+ *
+ * The corresponding classpath entry is not exposed by this interface. It's tied
+ * to it by a specific class that implements this interface.
+ */
+public interface DefinesClass {
+    /**
+     * Returns true if the classpath entry contains the requested class.
+     */
+    boolean apply(String className);
```
Comment (stuhood) [inline internal/compiler-interface/src/main/java/xsbti/compile/DefinesClass.java:13]: Boolean isn't the best API for this, as confirming that something is defined will usually take a similar amount of time to just fetching it (similar to `Map.get` vs `Map.contains+Map.apply`).

Would prefer to see `Maybe`/`Option` here, or removal of this API.

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala [hunk 2]
```diff
@@ -50,32 +51,38 @@ object Locate {
...
-      }
-      val entries = try { jar.entries.map(e => toClassName(e.getName)).toSet } finally { jar.close() }
-      entries.contains _
+  private object FalseDefinesClass extends DefinesClass {
+    override def apply(binaryClassName: String): Boolean = false
+  }
+
+  private class JarDefinesClass(entry: File) extends DefinesClass {
+    import collection.JavaConversions._
+    private val jar = try { new ZipFile(entry, ZipFile.OPEN_READ) } catch {
```
Comment (stuhood) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala:68]: AFAICT, this will unnecessarily hold a reference to the ZipFile in memory.

Maybe:

```
private val entries = {
  val jar = ???
  try { jar.entries... }
}
```

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala [hunk 3]
```diff
@@ -50,32 +51,38 @@ object Locate {
...
 
   def toClassName(entry: String): String =
     entry.stripSuffix(ClassExt).replace('/', '.')
 
   val ClassExt = ".class"
 
-  def directoryDefinesClass(entry: File): String => Boolean =
-    className => classFile(entry, className).isFile
+  private class DirectoryDefinesClass(entry: File) extends DefinesClass {
+    override def apply(binaryClassName: String): Boolean = classFile(entry, binaryClassName).isFile
```
Comment (stuhood) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala:84]: While you're in here...

the recursive creation of files in `def subDirectory` is likely causing filesystem operations, as each `new File` call executes normalization steps.

File: internal/compiler-interface/src/main/java/xsbti/compile/DefinesClass.java (hunk 1 above)
Comment (eed3si9n) [inline internal/compiler-interface/src/main/java/xsbti/compile/DefinesClass.java:13]: Since the calling side looks like this,

``` scala
val entries = classpath.toStream.map { entry => (entry, lookup.definesClass(entry)) }
```

and if there's frequent enough call, it might make sense to avoid instantiation of `Maybe`s.
If we're going to change it, I would either rename `apply` to some verb phrase or use [`xsbti.F1`](https://github.com/sbt/util/blob/1.0/internal/util-interface/src/main/java/xsbti/F1.java).

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala (hunk 2 above)
Comment (eed3si9n) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala:68]: I like the suggestion. We shouldn't need to hold on to zip file since the entries are converted to `Set[String]` anyway right?

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala (hunk 3 above)
Comment (eed3si9n) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/Locate.scala:84]: @stuhood Do you have a suggestion for improving `subDirectory`? If so, could you send a separate PR or post to sbt-dev to discuss it?

**Comment by eed3si9n** [comment]
This PR has been open for a while, so before we lose it I rebased against 1.0 and opened a new PR - https://github.com/sbt/zinc/pull/130, which I intend to merge when Travis passes.
